File: scripts/sdk/quickstarts/typescript/quickstart.mjs
// SPDX-License-Identifier: Apache-2.0

import { isAbsolute, relative, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { XenoteerClient, XenoteerError } from '@xenoteer/sdk';

function required(name) {
  const value = process.env[name];
  if (typeof value !== 'string' || value.length === 0)
    throw new Error(`required environment is missing: ${name}`);

  return value;
}

function verifyInstalledOrigin() {
  const expected = resolve(required('XENOTEER_EXPECTED_INSTALL_ROOT'));
  const installed = resolve(fileURLToPath(import.meta.resolve('@xenoteer/sdk')));
  const within = relative(expected, installed);
  if (within === '' || (!within.startsWith('..') && !isAbsolute(within)))
    return;

  throw new Error('TypeScript SDK resolved outside the staged npm installation');
}

async function exercise() {
  verifyInstalledOrigin();
  const language = required('XENOTEER_QUICKSTART_LANGUAGE');
  if (!/^[a-z-]+$/u.test(language))
    throw new Error('quick-start language label is invalid');

  const expectAuthenticationFailure = required('XENOTEER_EXPECT_AUTH_FAILURE') === '1';
  let client;
  try {
    client = await XenoteerClient.connect({
      baseUrl: required('XENOTEER_API_BASE'),
      token: required('XENOTEER_TOKEN'),
      requestTimeoutMs: 5_000,
    });
  } catch (error) {
    if (
      expectAuthenticationFailure
      && error instanceof XenoteerError
      && error.code === 'authentication'
      && error.status === 401
    ) {
      console.log(`quickstart-ok language=${language} mode=auth-failure`);
      return;
    }
    throw error;
  }

  try {
    if (expectAuthenticationFailure)
      throw new Error('invalid bearer unexpectedly authenticated');

    if (
      client.negotiatedProtocol.major !== 1
      || client.negotiatedProtocol.minor !== 0
    )
      throw new Error('server did not negotiate frozen protocol v1.0');

    if (client.status.desktop.state !== 'ready')
      throw new Error('desktop was not ready');

    client.desktop();
    console.log(`quickstart-ok language=${language} mode=success`);
  } finally {
    await client.close();
  }
}

try {
  await exercise();
} catch (error) {
  const detail = error instanceof XenoteerError
    ? `${error.name}[${error.code}]`
    : error instanceof Error
      ? error.message
      : 'unknown safe failure';
  console.error(`public TypeScript quick-start failed: ${detail}`);
  process.exitCode = 1;
}

File: scripts/sdk/quickstarts/python/quickstart.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""Minimal installed-package Xenoteer status quick-start."""

from __future__ import annotations

import asyncio
import os
import re
import sys
from pathlib import Path

import xenoteer
from xenoteer import ClientOptions, XenoteerClient, XenoteerError


def required(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"required environment is missing: {name}")
    return value


def verify_installed_origin() -> None:
    expected = Path(required("XENOTEER_EXPECTED_INSTALL_ROOT")).resolve(strict=True)
    module_file = Path(xenoteer.__file__).resolve(strict=True)
    if not module_file.is_relative_to(expected):
        raise RuntimeError("Python SDK resolved outside the staged distribution installation")


async def exercise() -> None:
    verify_installed_origin()
    language = required("XENOTEER_QUICKSTART_LANGUAGE")
    if re.fullmatch(r"[a-z-]+", language) is None:
        raise RuntimeError("quick-start language label is invalid")
    expect_authentication_failure = required("XENOTEER_EXPECT_AUTH_FAILURE") == "1"
    options = ClientOptions(
        base_url=required("XENOTEER_API_BASE"),
        token=required("XENOTEER_TOKEN"),
        request_timeout=5,
    )
    try:
        client = await XenoteerClient.connect(options)
    except XenoteerError as error:
        if (
            expect_authentication_failure
            and error.code == "authentication"
            and error.status == 401
        ):
            print(f"quickstart-ok language={language} mode=auth-failure")
            return
        raise

    async with client:
        if expect_authentication_failure:
            raise RuntimeError("invalid bearer unexpectedly authenticated")
        if (
            client.negotiated_protocol.major != 1
            or client.negotiated_protocol.minor != 0
        ):
            raise RuntimeError("server did not negotiate frozen protocol v1.0")
        if client.status.desktop.state != "ready":
            raise RuntimeError("desktop was not ready")
        client.desktop()
        print(f"quickstart-ok language={language} mode=success")


def main() -> int:
    try:
        asyncio.run(asyncio.wait_for(exercise(), timeout=8))
    except XenoteerError as error:
        print(
            f"public Python quick-start failed: XenoteerError[{error.code}]",
            file=sys.stderr,
        )
        return 1
    except (OSError, RuntimeError, asyncio.TimeoutError) as error:
        print(f"public Python quick-start failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
